In [10]:
#Install Goose3
#Install Sentence_transformers

from sentence_transformers import SentenceTransformer
import re
import nltk
import string
import numpy as np
import networkx as nx
from nltk.cluster.util import cosine_distance
import scipy
from sentence_transformers import SentenceTransformer, util
from goose3 import Goose
import pandas as pd

## Function to compare 2 article pages (e.g Url1 vs Url2)

In [11]:
url1 = 'https://fraserviewvet.ca/your-pet-should-be-doing-these-daily-exercises-to-stay-healthy/' # insert the first URL
url2 = 'https://fraserviewvet.ca/5-pet-hazards-to-be-aware-of-this-holiday-season/' # Insert the second URL

In [12]:
#Comparing just two URLS
def Two_URL_text_comparison (url1,url2):
    model = SentenceTransformer('bert-base-nli-mean-tokens')
    g = Goose()
    url1 = url1
    article1 = g.extract(url1)
    url2 = url2
    article2 = g.extract(url2)
    sentence1_representation = model.encode(article1.cleaned_text.replace("\n"," "))
    sentence2_representation = model.encode(article2.cleaned_text.replace("\n"," "))
    cosine_sim = util.pytorch_cos_sim(sentence1_representation,sentence2_representation)
    result = 'The cosine_sim for : ' + url1 + ' and ' + url2+ ' ' + 'is' + " = " + str(cosine_sim)
    print(result)
    return cosine_sim

# Compare Two Pages


In [13]:
cosine_sim = Two_URL_text_comparison(url1,url2) #run the model

The cosine_sim for : https://fraserviewvet.ca/your-pet-should-be-doing-these-daily-exercises-to-stay-healthy/ and https://fraserviewvet.ca/5-pet-hazards-to-be-aware-of-this-holiday-season/ is = tensor([[0.5772]])


Base on BERT, cosine similarity between two articles is 0.5772

## Comparing Multiple Pages

In [14]:
# function to compare multiple URLs
def multible_url_comparison (lsit_of_urls):
    result_data = pd.DataFrame(index = lsit_of_urls)
    model = SentenceTransformer('bert-base-nli-mean-tokens')
    g = Goose()
    for url1 in lsit_of_urls:
        url_cosine_sim = list()
        for url2 in lsit_of_urls:
            article1 = g.extract(url1)
            article2 = g.extract(url2)
            sentence1_representation = model.encode(article1.cleaned_text.replace("\n"," "))
            sentence2_representation = model.encode(article2.cleaned_text.replace("\n"," "))
            cosine_sim = util.pytorch_cos_sim(sentence1_representation,sentence2_representation)
            url_cosine_sim.append(cosine_sim)
        result_data[url1]= url_cosine_sim
        result_data_array = result_data.to_numpy()
    return result_data 

In [15]:
#insert a list of Url's for Comparison
lsit_of_urls = ['https://fraserviewvet.ca/5-pet-hazards-to-be-aware-of-this-holiday-season/',
                'https://fraserviewvet.ca/ways-to-keep-your-pet-entertained-this-winter/',
                'https://fraserviewvet.ca/3-reasons-why-regular-pet-grooming-is-essential/',
                'https://fraserviewvet.ca/follow-these-steps-to-ensure-pet-obesity-is-not-a-problem-for-your-pet/',
                'https://fraserviewvet.ca/5-signs-your-pet-is-in-pain/',
                'https://fraserviewvet.ca/suggestions-for-keeping-your-deaf-dog-safe/']

In [16]:
result_data = multible_url_comparison(lsit_of_urls) #run the model
result_data

https://fraserviewvet.ca/5-pet-hazards-to-be-aware-of-this-holiday-season/  \
https://fraserviewvet.ca/5-pet-hazards-to-be-aw...                                 [[tensor(1.0000)]]                           
https://fraserviewvet.ca/ways-to-keep-your-pet-...                                 [[tensor(0.7880)]]                           
https://fraserviewvet.ca/3-reasons-why-regular-...                                 [[tensor(0.7372)]]                           
https://fraserviewvet.ca/follow-these-steps-to-...                                 [[tensor(0.7354)]]                           
https://fraserviewvet.ca/5-signs-your-pet-is-in...                                 [[tensor(0.7650)]]                           
https://fraserviewvet.ca/suggestions-for-keepin...                                 [[tensor(0.6969)]]                           

                                                   https://fraserviewvet.ca/ways-to-keep-your-pet-entertained-this-winter/  \
https://fraserviewvet.ca/5-pet-hazards-to-be-aw...                                 [[tensor(0.7880)]]                        
https://fraserviewvet.ca/ways-to-keep-your-pet-...                                 [[tensor(1.0000)]]                        
https://fraserviewvet.ca/3-reasons-why-regular-...                                 [[tensor(0.8075)]]                        
https://fraserviewvet.ca/follow-these-steps-to-...                                 [[tensor(0.6539)]]                        
https://fraserviewvet.ca/5-signs-your-pet-is-in...                                 [[tensor(0.7017)]]                        
https://fraserviewvet.ca/suggestions-for-keepin...                                 [[tensor(0.6833)]]                        

                                                   https://fraserviewvet.ca/3-reasons-why-regular-pet-grooming-is-essential/  \
https://fraserviewvet.ca/5-pet-hazards-to-be-aw...                                 [[tensor(0.7372)]]                          
https://fraserviewvet.ca/ways-to-keep-your-pet-...                                 [[tensor(0.8075)]]                          
https://fraserviewvet.ca/3-reasons-why-regular-...                                 [[tensor(1.0000)]]                          
https://fraserviewvet.ca/follow-these-steps-to-...                                 [[tensor(0.7384)]]                          
https://fraserviewvet.ca/5-signs-your-pet-is-in...                                 [[tensor(0.8284)]]                          
https://fraserviewvet.ca/suggestions-for-keepin...                                 [[tensor(0.7534)]]                          

                                                   https://fraserviewvet.ca/follow-these-steps-to-ensure-pet-obesity-is-not-a-problem-for-your-pet/  \
https://fraserviewvet.ca/5-pet-hazards-to-be-aw...                                 [[tensor(0.7354)]]                                                 
https://fraserviewvet.ca/ways-to-keep-your-pet-...                                 [[tensor(0.6539)]]                                                 
https://fraserviewvet.ca/3-reasons-why-regular-...                                 [[tensor(0.7384)]]                                                 
https://fraserviewvet.ca/follow-these-steps-to-...                                 [[tensor(1.0000)]]                                                 
https://fraserviewvet.ca/5-signs-your-pet-is-in...                                 [[tensor(0.7498)]]                                                 
https://fraserviewvet.ca/suggestions-for-keepin...                                 [[tensor(0.7317)]]                                                 

                                                   https://fraserviewvet.ca/5-signs-your-pet-is-in-pain/  \
https://fraserviewvet.ca/5-pet-hazards-to-be-aw...                                 [[tensor(0.7650)]]      
https://fraserviewvet.ca/ways-to-keep-your-pet-...                                 [[tensor(0